### PR TITLE
refactored kafka consumer

### PR DESCRIPTION
**Pull request by fabregas**

### CODE REVIEW

File: aiokafka/consumer.py [hunk 1]
```diff
@@ -1,375 +1,623 @@
...
+            return
+        log.debug("Closing the KafkaConsumer.")
+        self._closed = True
+        if self._coordinator:
+            yield from self._coordinator.close()
+        if self._fetcher:
+            yield from self._fetcher.close()
+        self._client.close()
+        try:
+            self._key_deserializer.close()
```
Comment (tvoinarovskyi) [inline aiokafka/consumer.py]: We documented, that deserializer is a callable. Nothing about close method there. Maybe it's some backward-compatibility for python-kafka, and we don't need it?

File: aiokafka/consumer.py [hunk 2]
```diff
@@ -1,375 +1,623 @@
...
+        except AttributeError:
+            pass
+        try:
+            self._value_deserializer.close()
```
Comment (tvoinarovskyi) [inline aiokafka/consumer.py]: same

File: aiokafka/consumer.py [hunk 3]
```diff
@@ -1,375 +1,623 @@
...
+        """Close the consumer, waiting indefinitely for any needed cleanup."""
+        if self._closed:
+            return
+        log.debug("Closing the KafkaConsumer.")
+        self._closed = True
+        if self._coordinator:
+            yield from self._coordinator.close()
+        if self._fetcher:
+            yield from self._fetcher.close()
+        self._client.close()
```
Comment (tvoinarovskyi) [inline aiokafka/consumer.py]: I think this is a coroutine. Only connection's `close` is a function

File: aiokafka/consumer.py [hunk 4]
```diff
@@ -1,375 +1,623 @@
...
+        Currently only supports kafka-topic offset storage (not zookeeper)
 
-        self._partitions = partitions
-        self._commit_task = None
+        Arguments:
+            offsets (dict, optional): {TopicPartition: OffsetAndMetadata} dict
+                to commit with the configured group_id. Defaults to current
+                consumed offsets for all subscribed partitions.
+        """
+        assert self._group_id is not None, 'Requires group_id'
```
Comment (tvoinarovskyi) [inline aiokafka/consumer.py:293]: As I see you removed the asserts of `python-kafka`, as they are inside coordinator. I think it's better to do it here

Comment (tvoinarovskyi) [inline aiokafka/consumer.py:293]: How do you think, should we allow to save offsets, that are not part of our assigned partitions?

File: aiokafka/consumer.py [hunk 5]
```diff
@@ -1,375 +1,623 @@
...
-        partitions = self._client.get_partition_ids_for_topic(self._topic)
-        for partition in partitions:
-            req = OffsetFetchRequest(self._topic, partition)
-            try:
-                (resp,) = yield from self._client.send_offset_fetch_request(
-                    self._group, [req])
-                partition_offset = resp.offset
-            except UnknownTopicOrPartitionError:
-                partition_offset = 0
+    def partitions_for_topic(self, topic):
```
Comment (tvoinarovskyi) [inline aiokafka/consumer.py:346]: Should we leave this as a function? Maybe we might want to wait here?

File: aiokafka/consumer.py [hunk 6]
```diff
@@ -1,375 +1,623 @@
...
-                offset = self._offsets[partition]
-                log.debug("Commit offset %d in SimpleConsumer: "
-                          "group=%s, topic=%s, partition=%s" %
-                          (offset, self._group, self._topic, partition))
+    def pause(self, *partitions):
+        """Suspend fetching from the import requested partitions.
 
-                reqs.append(OffsetCommitRequest(self._topic, partition,
-                                                offset, None))
+        Future calls to poll() will not return any records from these
```
Comment (Drizzt1991) [inline aiokafka/consumer.py]: This comment is plain wrong. Does it even work?

File: aiokafka/consumer.py [hunk 7]
```diff
@@ -1,375 +1,623 @@
...
-            if not result:
-                continue
-            partition, message = result
-            if self._partition_info:
-                messages.append(result)
-            else:
-                messages.append(message)
-            new_offsets[partition] = message.offset + 1
-            count -= 1
+    def _init_fetcher(self):
```
Comment (tvoinarovskyi) [inline aiokafka/consumer.py]: Can't find where it's used. Ghoust?

File: aiokafka/consumer.py [hunk 8]
```diff
@@ -1,375 +1,623 @@
...
+        On each poll, consumer will try to use the last consumed offset as the
+        starting offset and fetch sequentially. The last consumed offset can be
+        manually set through seek(partition, offset) or automatically set as
+        the last committed offset for the subscribed list of partitions.
+        Incompatible with iterator interface (get_message) -- use one or the
+        other, not both.
+
+        Arguments:
+            timeout_ms (int, optional): milliseconds spent waiting in poll if
+                data is not available in the buffer. If 0, returns immediately
```
Comment (Drizzt1991) [inline aiokafka/consumer.py]: Mismatch with implementation. 0 can't be used

File: aiokafka/consumer.py [hunk 9]
```diff
@@ -1,375 +1,623 @@
...
+        start = self._loop.time()
+        # fetch positions if we have partitions we're subscribed
+        # to that we don't know the offset for
+        if not self._subscription.has_all_fetch_positions():
+            partitions = self._subscription.missing_fetch_positions()
+            yield from self._update_fetch_positions(partitions)
+
+        records = self._fetcher.fetched_records()
+        if not records:
+            self._fetcher.init_fetches()
```
Comment (tvoinarovskyi) [inline aiokafka/consumer.py]: it's awaitable, why do we wait in a loop?

File: aiokafka/fetcher.py [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+            try:
+                return next(self._iterator)
+            except StopIteration:
+                self._iterator = None
+                continue
+
+    def fetched_records(self):
+        """Returns previously fetched records and updates consumed offsets."""
+        drained = collections.defaultdict(list)
+        for message in self._message_generator():
```
Comment (tvoinarovskyi) [inline aiokafka/fetcher.py]: This will have alot of overhead... Why is it needed?

File: aiokafka/fetcher.py [hunk 11]
```diff
@@ -0,0 +1,567 @@
...
+        self._client = client
+        self._loop = loop
+        self._key_deserializer = key_deserializer
+        self._value_deserializer = value_deserializer
+        self._fetch_min_bytes = fetch_min_bytes
+        self._fetch_max_wait_ms = fetch_max_wait_ms
+        self._max_partition_fetch_bytes = max_partition_fetch_bytes
+        self._check_crcs = check_crcs
+        self._subscriptions = subscriptions
+        self._records = asyncio.Queue(loop=loop)
```
Comment (tvoinarovskyi) [inline aiokafka/fetcher.py]: Why do you use `asyncio.Queue` if you never use blocking methods on it?

File: aiokafka/consumer.py (hunk 4 above)
Comment (fabregas) [inline aiokafka/consumer.py:293]: I think consumer can commit any partition in any topic

File: aiokafka/fetcher.py (hunk 10 above)
Comment (fabregas) [inline aiokafka/fetcher.py]: what do you mean? we parse messages in generator and "save" it in dict{topicPartition: list of messages} .. 
what do you suggest?

**Comment by TarasLevelUp** [comment]
Lets move this to an official PR

**Pull request merged**